### PR TITLE
Added timing probes

### PR DESCRIPTION
**Pull request by anki-code**
Hi! 

1. I've found that PygmentsLexer takes much time on first start and I add timing probe to show it.
2. I've found that `prompt_tokens` takes much time and I add timing probe to show it.
3. Added `on_before_prompter_prompt` event to show timings in the right place.

Example of no-rc start:
```
xonsh --timings --no-rc
|-------------------------------|-----------|-----------|
|Event name                     | Time (s)  | Delta (s) |
|-------------------------------|-----------|-----------|
|start                          |   0.000   |   0.000   |
|pre_execer_init                |   0.016   |   0.016   |
|post_execer_init               |   0.064   |   0.048   |
|on_pre_rc                      |   0.064   |   0.000   |
|on_post_rc                     |   0.064   |   0.000   |
|on_ptk_create                  |   0.145   |   0.081   |
|on_post_init                   |   0.145   |   0.000   |
|on_pre_cmdloop                 |   0.213   |   0.068   |
|on_prompt_singleline_create    |   0.213   |   0.000   |
|on_pre_prompt_tokens           |   0.213   |   0.000   |
|on_post_prompt_tokens          |   0.543   |   0.330   | <===========================
|on_pre_prompt_pygments_lexer   |   0.543   |   0.000   |
|on_post_prompt_pygments_lexer  |   0.585   |   0.041   |
|on_before_prompter_prompt      |   0.585   |   0.000   |
|-------------------------------|-----------|-----------|
```

Example of my rc start:
```
$ xonsh --timings 
|-------------------------------|-----------|-----------|
|Event name                     | Time (s)  | Delta (s) |
|-------------------------------|-----------|-----------|
|start                          |   0.000   |   0.000   |
|pre_execer_init                |   0.015   |   0.015   |
|post_execer_init               |   0.061   |   0.046   |
|on_pre_rc                      |   0.061   |   0.000   |
|on_post_rc                     |   0.486   |   0.425   |
|on_ptk_create                  |   0.505   |   0.019   |
|on_post_init                   |   0.506   |   0.000   |
|on_pre_cmdloop                 |   0.506   |   0.000   |
|on_prompt_singleline_create    |   0.506   |   0.000   |
|on_pre_prompt_tokens           |   0.534   |   0.028   |
|on_post_prompt_tokens          |   0.544   |   0.010   |
|on_pre_prompt_pygments_lexer   |   0.544   |   0.000   |
|on_post_prompt_pygments_lexer  |   0.879   |   0.335   | <==========================
|on_before_prompter_prompt      |   0.879   |   0.000   |
|-------------------------------|-----------|-----------|
```

### CODE REVIEW

File: xonsh/ptk_shell/shell.py [hunk 1]
```diff
@@ -219,8 +222,8 @@ def singleline(
             # once the prompt is done, update it in background as each future is completed
             prompt_args["pre_run"] = self.prompt_formatter.start_update
 
+        events.on_pre_prompt.fire()
```
Comment (laloch) [inline xonsh/ptk_shell/shell.py:225]: I don't think it's ok to move `on_pre_prompt` event this far down. This way the event handler doesn't have any control over `prompt_args` anymore.

Comment (anki-code) [reply xonsh/ptk_shell/shell.py:225]: What is "control over prompt_args"?

Comment (laloch) [reply xonsh/ptk_shell/shell.py:228]: The whole point of `pre_XYZ` events is to give the user control over what `XYZ` will do. For instance, the point of `pre_render` event would be to give the user the last chance to influence how the subject of the rendering is going to be rendered via event handler. It's a paradigm/pattern. This change moves the event to the point where the event handler can't influence the outcome of the respective operation (prompt rendering) at all.

Comment (anki-code) [reply xonsh/ptk_shell/shell.py:226]: Got it. I just forget about globals that could be changed via on_pre_prompt, you're right. Then I add new event immediately before show the prompt. Could you please help with the name?
```
        events.on_NEW_EVENT_NAME.fire()
        line = self.prompter.prompt(**prompt_args)
```

Comment (laloch) [reply xonsh/ptk_shell/shell.py:226]: Well, the name depends on what the point of the new event is. What's the point?

Comment (anki-code) [reply xonsh/ptk_shell/shell.py]: Added new `on_before_prompter_prompt` event. Changes in the name are welcome.

**Comment by anki-code** [comment]
Ready

File: xonsh/shell.py [hunk 2]
```diff
@@ -75,6 +75,15 @@
 """,
 )
 
+events.doc(
+    "on_before_prompter_prompt",
```
Comment (scopatz) [inline xonsh/shell.py:79]: probably should follow the pre/post naming convention, ie `on_pre_prompter_prompt`

**Comment by anki-code** [comment]
Ready

File: xonsh/ptk_shell/shell.py [hunk 3]
```diff
@@ -219,8 +223,8 @@ def singleline(
             # once the prompt is done, update it in background as each future is completed
             prompt_args["pre_run"] = self.prompt_formatter.start_update
 
+        events.on_pre_prompter_prompt.fire()
```
Comment (scopatz) [inline xonsh/ptk_shell/shell.py:226]: Doesn't this need to be added to the readline shell then too?

Comment (anki-code) [reply xonsh/ptk_shell/shell.py:226]: We're thinking similar and I already take a look at readline when worked on pr.
I really not the user of readline and have zero motivation to do it in readline. The code in the readline shell so different that requres to additional redesign to make probes similar with ptk. So I leave the idea of adding it to readline.

Comment (scopatz) [reply xonsh/ptk_shell/shell.py:226]: Well, this is more of a change to the API that about when the timings are reported. So if I understand this PR correctly, it breaks timings with readline.  This should be a simple one-line fix

Comment (anki-code) [reply xonsh/ptk_shell/shell.py:226]: done

**Comment by anki-code** [comment]
Ready

File: xonsh/readline_shell.py [hunk 4]
```diff
@@ -366,7 +366,9 @@ def singleline(self, store_in_history=True, **kwargs):
                 store_in_history = True
             pos = readline.get_current_history_length() - 1
         events.on_pre_prompt.fire()
-        rtn = input(self.prompt)
+        prompt = self.prompt
```
Comment (scopatz) [inline xonsh/readline_shell.py:369]: you don't need this line.

Comment (scopatz) [reply xonsh/readline_shell.py:369]: Oh whoops sorry, you do

**Comment by scopatz** [comment]
Thanks @anki-code! I think I see what you are trying to do a little better here now.  I would like to recommend some name changes that I think are more descriptive and correct.

```python
on_pre_prompt -> on_pre_prompt_format
on_pre_prompt_tokens -> on_pre_prompt_tokenize  # make a verb
on_post_prompt_tokens -> on_post_prompt_tokenize  # make a verb
on_pre_prompt_pygments_lexer -> on_pre_prompt_style
on_post_prompt_pygments_lexer -> on_post_prompt_style
on_pre_prompter_prompt -> on_pre_prompt  # return to the original name
```

File: xonsh/ptk_shell/shell.py [hunk 5]
```diff
@@ -198,10 +199,13 @@ def singleline(
             "refresh_interval": refresh_interval,
             "complete_in_thread": complete_in_thread,
         }
+
         if env.get("COLOR_INPUT"):
             if HAS_PYGMENTS:
+                events.on_timingprobe.fire(name="on_pre_prompt_pygments_lexer")
```
Comment (scopatz) [inline xonsh/ptk_shell/shell.py:205]: This should be moved up one line

File: xonsh/ptk_shell/shell.py [hunk 6]
```diff
@@ -198,10 +199,13 @@ def singleline(
...
                 prompt_args["lexer"] = PygmentsLexer(pyghooks.XonshLexer)
                 style = style_from_pygments_cls(pyghooks.xonsh_style_proxy(self.styler))
+                events.on_timingprobe.fire(name="on_post_prompt_pygments_lexer")
```
Comment (scopatz) [inline xonsh/ptk_shell/shell.py:208]: this should be moved below the else-clause

**Comment by anki-code** [comment]
Ready

**Comment by scopatz** [comment]
Thanks @anki-code! This looks good to me, and always good to have more timings!

**Comment by anki-code** [comment]
Yeah ) Now you can go to the main dish on the table - #3917 )))

**Pull request merged**